### PR TITLE
ngtcp2-gnutls: init at 0.7.0 and use in knot-dns

### PR DESCRIPTION
**Pull request by vcunat**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
@Izorkin might want to know about the split or comment.  Anyway, I don't expect significant usage of the openssl variant before the official openssl is made compatible with QUIC.

**Comment by Izorkin** [comment]
Not working this variant?
```diff
index 90372cc97a2..8c75eb1b8c5 100644
--- a/pkgs/development/libraries/ngtcp2/default.nix
+++ b/pkgs/development/libraries/ngtcp2/default.nix
@@ -1,10 +1,14 @@
 { lib, stdenv, fetchFromGitHub
 , autoreconfHook, pkg-config, file
-, libev, nghttp3, quictls
+, libev, nghttp3
 , cunit, ncurses
 , withJemalloc ? false, jemalloc
+, withGnutls ? false, gnutls
+, withQuictls ? true, quictls
 }:

+assert !(withGnutls && withQuictls);
+
 stdenv.mkDerivation rec {
   pname = "ngtcp2";
   version = "0.7.0";
@@ -19,9 +23,16 @@ stdenv.mkDerivation rec {
   outputs = [ "out" "dev" "doc" ];

   nativeBuildInputs = [ autoreconfHook pkg-config file ];
-  buildInputs = [ libev nghttp3 quictls ] ++ lib.optional withJemalloc jemalloc;
+
+  buildInputs = [ libev nghttp3 ]
+    ++ lib.optional withQuictls quictls
+    ++ lib.optional withGnutls gnutls
+    ++ lib.optional withJemalloc jemalloc;
+
   checkInputs = [ cunit ncurses ];

+  configureFlags = lib.optional withGnutls "--with-gnutls=yes";
+
   preConfigure = ''
     substituteInPlace ./configure --replace /usr/bin/file ${file}/bin/file
   '';
diff --git a/pkgs/servers/dns/knot-dns/default.nix b/pkgs/servers/dns/knot-dns/default.nix
index 23fefe80e76..fff31f86df3 100644
--- a/pkgs/servers/dns/knot-dns/default.nix
+++ b/pkgs/servers/dns/knot-dns/default.nix
@@ -1,5 +1,5 @@
 { lib, stdenv, fetchurl, pkg-config, gnutls, liburcu, lmdb, libcap_ng, libidn2, libunistring
-, systemd, nettle, libedit, zlib, libiconv, libintl, libmaxminddb, libbpf, nghttp2, libmnl
+, systemd, nettle, libedit, zlib, libiconv, libintl, libmaxminddb, libbpf, ngtcp2, nghttp2, libmnl
 , autoreconfHook, nixosTests, knot-resolver
 }:

@@ -32,6 +32,7 @@ stdenv.mkDerivation rec {
     gnutls liburcu libidn2 libunistring
     nettle libedit
     libiconv lmdb libintl
+    ngtcp2
     nghttp2 # DoH support in kdig
     libmaxminddb # optional for geoip module (it's tiny)
     # without sphinx &al. for developer documentation
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index bb0430ef58d..6d7576bacef 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22657,7 +22657,10 @@ with pkgs;

   keycloak = callPackage ../servers/keycloak { };

-  knot-dns = callPackage ../servers/dns/knot-dns { };
+  knot-dns = callPackage ../servers/dns/knot-dns {
+    ngtcp2 = ngtcp2.override { withGnutls = true; withQuictls = false; };
+  };
+
   knot-resolver = callPackage ../servers/dns/knot-resolver {
     systemd = systemdMinimal; # in closure already anyway
   };
```

**Comment by vcunat** [comment]
Yes, it's certainly possible to merge this way.  But I might also need ngtcp2 version kept in lock-step with knot-dns.  Until nghtcp2 development stabilizes.  So the overal code sharing opportunities don't seem significant.

**Comment by Izorkin** [comment]
Then it will be easier to have two variants.
Knot-dns can't work with quictls?

**Comment by vcunat** [comment]
No, it only uses gnutls for crypto.  (including parts that are not related to TLS or QUIC)

**Review by Izorkin — APPROVED** [review]

File: pkgs/development/libraries/ngtcp2/gnutls.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    platforms = platforms.unix;
+    maintainers = with maintainers; [ vcunat/* for knot-dns */ ];
+  };
+}
+
+/*
+  Why split from ./default.nix?
+
+  ngtcp2 libs contain helpers to plug into various crypto libs (gnutls, patched openssl, ...).
+  Building multiple of them while keeping closures separable would be relatively complicated.
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/ngtcp2/gnutls.nix:43]: Would it be a problem to compile openssl and gnutls into one package?

Comment (vcunat) [reply pkgs/development/libraries/ngtcp2/gnutls.nix:43]: You would need to put each of these into a separate output (and leaving the base lib in a third one) &ndash; and then patch paths in pkg-config and libtool files.

Comment (vcunat) [reply pkgs/development/libraries/ngtcp2/gnutls.nix:43]: No, that's a different one.  The openssl fork takes a few megabytes and gnutls also (and each additional crypto lib that you'd add).  Packages always want just one but they'd get all of them.

Comment (vcunat) [reply pkgs/development/libraries/ngtcp2/gnutls.nix:43]: The potential blowup in this comment is for the situation that got merged.  That is you maybe need (through different packages) ngtcp2 with both openssl and gnutls, in which case you get the core twice.  And the core is 0.3--0.4 MB.

**Pull request merged**